Feature: 이벤트 팝업 & 배너 구현

inwoodev opened the pull request:
## PR 타입
어떤 변경에 대한 PR인가요?

<!-- 아래 체크리스트 중 해당 되는 부분에 체크해주세요 "x". -->

- [ ] 버그 수정
- [x] Feature / 신기능
- [ ] 코드 스타일 업데이트 (formatting, local variables)
- [ ] 리팩토링 (no functional changes, no api changes)
- [ ] 빌드 관련 변경사항
- [ ] CI 관련된 변경사항
- [ ] Documentation content changes
- [ ] 구조 변경
- [ ] Other... 추가 설명 요망:


## 배경
- Feelin 이벤트 발생 시 관련 팝업을 유저에게 보여줄 수 있어야 하며 홈화면 상단 배너 또한 관련 이벤트에 맞게 변경되어야 합니다.

## 목표
- 이벤트 존재 시 홈화면에서 이벤트 팝업 출력 (로그인 & 비로그인 유저 모두 해당)
- 이벤트 존재 시 홈화면 상단 배너에서 관련 내용 출력 (로그인 & 비로그인 유저 모두 해당)
- 이밴트 거부 시 하루동안 팝업이 뜨지 않도록 설정

## 결과 (Optional)
- [x] 이벤트 팝업 출력 기능 구현
- [x] 이벤트 배너 적용

|이벤트 팝업(로그인)|이벤트 팝업(비로그인)|
|--|--|
|<img src="https://github.com/user-attachments/assets/42c47cc3-6628-4dc1-88a1-296dc4de6208" width="250">|<img src="https://github.com/user-attachments/assets/191336e0-8cdf-4902-8597-c3f35196e8ac" width="250">|

inwoodev committed:
feat: PopUpViewController & FeelinEventView 구현

inwoodev committed:
feat: 홈화면에서 이벤트 팝업을 보여주는 기능 구현

inwoodev committed:
feat : 이벤트 배너 API & 적용

inwoodev commented on Projects/Shared/DesignSystem/Sources/Extensions/UIViewController+.swift:
```diff
@@ -191,3 +191,73 @@ public extension UIViewController {
+        present(popUpViewController, animated: false)
+    }
+    
+    public func showEventPopUp(
+        contentImageUrl: URL,
+        leftTopActionTitle: String = "오늘 하루 보지 않기",
+        rightTopActionImage: UIImage = FeelinImages.x.withTintColor(.white),
+        leftTopActionCompletion: (() -> Void)?,
+        rightTopActionCompletion: (() -> Void)?,
+        bottomActionTitle: String,
+        bottomActionCompletion: (() -> Void)?
+    ) {
+        self.showPopUp(
+            contentImageUrl: contentImageUrl,
+            leftTopActionTitle: leftTopActionTitle,
+            leftTopActionImage: nil,
+            rightTopActionTitle: nil,
+            rightTopActionImage: rightTopActionImage,
+            leftTopActionCompletion: leftTopActionCompletion,
+            rightTopActionCompletion: rightTopActionCompletion,
+            bottomActionTitle: bottomActionTitle,
+            bottomActionCompletion: bottomActionCompletion
+        )
+    }
```
기본적으로 해당 메서드를 활용하여 이벤트 팝업 alert를 화면에 띄울 수 있습니다.

inwoodev commented on Projects/Shared/DesignSystem/Sources/Views/FeelinEventView.swift:
```diff
@@ -0,0 +1,158 @@
+            activityIndicator.centerXAnchor.constraint(equalTo: eventImageView.centerXAnchor),
+            activityIndicator.centerYAnchor.constraint(equalTo: eventImageView.centerYAnchor),
+            
+            stackView.topAnchor.constraint(equalTo: self.topAnchor),
+            stackView.leadingAnchor.constraint(equalTo: self.leadingAnchor),
+            stackView.trailingAnchor.constraint(equalTo: self.trailingAnchor),
+            stackView.bottomAnchor.constraint(equalTo: self.feedbackButton.topAnchor),
```
단순 이미지 뷰를 활용할 수도 있었겠지만 추후 단순이 이미지 뿐 아니라 다른 요소들이 들어가게 될 때 더 범용적이게 활용할 수 있을 듯 하여 stackView에 이미지뷰를 넣는 방식으로 오토레이아웃을 잡았습니다.

inwoodev commented on Projects/Shared/DesignSystem/Sources/Views/FeelinEventView.swift:
```diff
@@ -0,0 +1,158 @@
+        ])
+    }
+    
+    public func loadImage(
+        from url: URL,
+        retryCount: Int = 3
+    ) {
+        DispatchQueue.main.async {
+            self.activityIndicator.startAnimating()
+        }
+
+        URLSession.shared.dataTask(with: url) { data, response, error in
+            DispatchQueue.main.async {
+                if let image = self.processImageData(data, error: error) {
+                    self.eventImageView.image = image
+                    self.activityIndicator.stopAnimating()
+                } else if retryCount > 0 {
+                    self.retryLoadingImage(from: url, retryCount: retryCount - 1)
+                }
+            }
+        }.resume()
+    }
```
이미지 뷰가 로딩되는 시간 동안 activityIndicator를 화면에 띄우기 위해 이와 같이 loadImage 메서드를 추가하였습니다.
retryCount기능 또한 일단 넣어서 실패할 경우 커스텀하게 이미지 요청을 재시도할 수 있도록 설계하였습니다~!

inwoodev commented on Projects/Shared/DesignSystem/Sources/Views/FeelinEventView.swift:
```diff
@@ -0,0 +1,158 @@
+        }
+    }
+    
+    private func adjustImageHeight() {
+        let imageViewWidth: CGFloat = self.frame.width
+        
+        let defaultAspectRatio: CGFloat = 1.25
+        
+        var calculatedHeight = imageViewWidth * defaultAspectRatio
+        
+        // 새로운 높이로 업데이트
+        imageHeightConstraint = eventImageView.heightAnchor.constraint(equalToConstant: calculatedHeight)
+        imageHeightConstraint.isActive = true
+    }
```
저희 피그마에 올라온 사이즈를 보았을 때  Height/Width가 대략적으로 1.25의 비율이 되더라구요.
그래서 비동기적으로 받아온 이미지를 적용한 뒤 이미지의 heightAnchor를 변경하는 방식으로 이미지 크기를 조정하였습니다.

inwoodev committed:
refactor: 배너 UI 업데이트 성능 문제 해결

inwoodev commented on Projects/Shared/DesignSystem/Sources/Views/FeelinEventView.swift:
```diff
@@ -0,0 +1,158 @@
+        setupView()
+    }
+    
+    public override func layoutSubviews() {
+        super.layoutSubviews()
+        // layoutSubviews에서 이미지 크기를 조정한다.
+        self.adjustImageHeight()
+    }
```
`layotuSubviews()`에서 `adjustImageHeight()`를 호출하지 않고

이미지를 적용하는 시점에 이와 같이 이미지 height설정을 둘 수도 있지만
``` swift
public func loadImage(
        from url: URL,
        retryCount: Int = 3
    ) {
        DispatchQueue.main.async {
            self.activityIndicator.startAnimating()
        }

        URLSession.shared.dataTask(with: url) { data, response, error in
            DispatchQueue.main.async {
                if let image = self.processImageData(data, error: error) {
                    self.eventImageView.image = image
                    self.activityIndicator.stopAnimating()
                    self.adjustImageHeight()
                } else if retryCount > 0 {
                    self.retryLoadingImage(from: url, retryCount: retryCount - 1)
                }
            }
        }.resume()
    }
```

![Simulator Screen Recording - iPhone 15 Pro - 2025-03-22 at 22 28 44](https://github.com/user-attachments/assets/1d373b88-a16f-41f2-87ac-58a7e0bafb6e)

그렇게 되면 이와 같이 팝업 크기가 갑자기 커져서 UX적으로 안좋다고 판단이 들어서 애초에 뷰의 크기가 정해지는 시점에 빈 이미지뷰라 하더라도 크기를 미리 조정 해 두는게 낫다고 판단이 들었기에 `layotuSubviews()`에서 `adjustImageHeight()`를 호출하도록 설계하였습니다.

inwoodev commented on Projects/Feature/Home/Interface/Sources/ViewControllers/HomeViewController.swift:
```diff
@@ -213,13 +213,22 @@ public class HomeViewController: UIViewController, NoteMenuHandling, NoteMusicHa
-        // 배너 섹션이 없는 경우 추가
-        if !snapshot.sectionIdentifiers.contains(.banner) {
+        
+        // 배너 섹션이 있는지 확인
+        if snapshot.sectionIdentifiers.contains(.banner) {
+            // 기존 아이템 삭제
+            let existingItems = snapshot.itemIdentifiers(inSection: .banner)
+            snapshot.deleteItems(existingItems)
+        } else {
+            // 섹션이 없으면 추가
             snapshot.appendSections([.banner])
         }
+        
+        // 새로운 아이템 추가
         snapshot.appendItems(newItems, toSection: .banner)
-        homeDataSource.applySnapshotUsingReloadData(snapshot)
+        
+        // 스냅샷 적용
```
이미 존재하는 아이템을 다시 insert할 경우 이와 같은 경고 메세지가 나오게 됩니다:

```
Diffable data source detected an attempt to insert or append 1 item identifier that already exists in the snapshot.
The existing item identifier will be moved into place instead, but this operation will be more expensive. 
For best performance, inserted item identifiers should always be unique. 
Set a symbolic breakpoint on BUG_IN_CLIENT_OF_DIFFABLE_DATA_SOURCE__IDENTIFIER_ALREADY_EXISTS
to catch this in the debugger. Item identifier that already exists: 
```
퍼포먼스 문제를 해결하기 위해 기존 아이템을 삭제한 뒤 다시 아이템을 적용하는 방식으로 로직을 수정하였습니다.

derrickkim0109 commented on Projects/Domain/Shared/Interface/Sources/Errors/EventError.swift:
```diff
@@ -0,0 +1,83 @@
+import Shared
+
+public enum EventError: LocalizedError, Equatable {
+    case feelinAPIError(FeelinAPIError)
+    case networkError(NetworkError)
+    case keychainError(KeychainError)
+    case unknown(errorDescription: String)
```
이 케이스들이 여러 군데에서 사용되고 있는데 나중에 해당 케이스들만 있는 상위 Protocol로 있으면 어떨까요? 
errorMessageWithCode, userMessage 정도만 해당 객체에서 선언해도 될 것 같다는 생각이듭니다. 
어떻게 생각하세요?

inwoodev replied:
음 일단 protocol 같은 경우 case들을 가질 수 없기 때문에 해당 케이스들을 상위 프로토콜로 빼는건 불가합니다!

공통 case를 갖고 있는 enum을 상속하는 것 또한 enum 특성상 불가하구요. 나중에 공통화를 한다면 다음과 같은 구조로 리팩토링해 볼 수 있을 듯 합니다.

1. FeelinCommonError와 같이 하나의 enum타입으로 통일화
- 기존 파편화 되어있는 NoteError, EventError들을 하나의 타입으로 통일

2. 프로토콜을 활용하여 에러 추상화
``` swift
public protocol FeelinCommonError: LocalizedError, Equatable {
    var errorMessage: String { get }
    var errorCode: String? { get }
    var errorMessageWithCode: String { get }
    var userMessage: String { get }
    var data: AnyType? { get }
}
```

요런 접근은 어떻게 생각하시나요~?

derrickkim0109 replied:
오우 좋습니다. 케이스는 상속이나 채택할 수 없다는 걸 깜빡했습니다.
제가 설명을 잘못 했는데 제 요지도 저런 부분이였습니다~!

inwoodev replied:
요건 추후 리팩토링 시 전반적으로 수정하시죠~~

derrickkim0109 replied:
넵 좋습니다.

derrickkim0109 commented on Projects/Domain/Shared/Interface/Sources/UseCasess/FetchSingleEventUseCase.swift:
```diff
@@ -0,0 +1,37 @@
+    public func execute() -> AnyPublisher<FeelinEvent, EventError> {
+        return eventAPIService.getEvents()
+            .compactMap { response in
+                response.event.data.first.map {
+                    FeelinEvent(
+                        dto: $0,
+                        refusalPeriod: response.refusalPeriod
+                    )
+                }
```
**P3**

왜 데이터의 첫 번째 값만 맵핑하는지 이유를 알 수 있을까요? 

주석으로 남겨주셔도 좋을 것 같습니다.

inwoodev replied:
주석 추가하겠습니다. 코멘트로도 말씀 드리면 usecase 네이밍과 같이 `FetchSingleEventUseCase`이기에 받아온 데이터 중 하나만 필요로 하기에 첫 번째 값만 매핑하였습니다. 서버로부터 데이터를 한 개만 받는 것이 확실하기에 last를 써도 무방합니다.

derrickkim0109 replied:
넵 이해했습니다~!

derrickkim0109 commented on Projects/Shared/DesignSystem/Sources/Views/FeelinEventView.swift:
```diff
@@ -0,0 +1,158 @@
+        setupView()
+    }
+    
+    required public init?(coder: NSCoder) {
+        super.init(coder: coder)
+        setupView()
+    }
```
해당 init에서도 호출될 일이 있을까요?

inwoodev replied:
불필요하겠네요. 수정하겠습니다 :)

derrickkim0109 replied:
감사합니다~!

derrickkim0109 reviewed (commented):
수고하셨습니다~! 
간단한 코멘트 남겼습니다.

inwoodev committed:
refactor: 주석 & 불필요한 코드 제거

- FetchSingleEventUseCase에 첫 번째 데이터만 매핑하는 이유 주석 추가
- FeelinEventView의 required init 비활성화

derrickkim0109 reviewed (approved):
수고하셨습니다~!
